### PR TITLE
adding possibility to set solver options for MOI constructor

### PR DESCRIPTION
**Pull request by vitornesello**
Answering issue https://github.com/JuliaOpt/Cbc.jl/issues/79

### CODE REVIEW

File: src/MOIWrapper.jl [hunk 1]
```diff
@@ -8,7 +8,15 @@ const CbcCI = CbcCInterface
 
 mutable struct Optimizer <: MOI.AbstractOptimizer
     inner::CbcModel
-    Optimizer() = new(CbcModel()) # Initializes with an empty model
+    Optimizer(::Nothing) = new(CbcModel()) # Initializes with an empty model
```
Comment (mlubin) [inline src/MOIWrapper.jl]: Why is the ``::Nothing`` needed?

File: src/MOIWrapper.jl [hunk 2]
```diff
@@ -8,7 +8,15 @@ const CbcCI = CbcCInterface
...
+end
+
+function Optimizer(;kwargs...)        
+    model = Optimizer(nothing)
+    for (name,value) in kwargs
```
Comment (mlubin) [inline src/MOIWrapper.jl]: Formatting nit: `(name, value)`

File: src/MOIWrapper.jl [hunk 3]
```diff
@@ -8,7 +8,15 @@ const CbcCI = CbcCInterface
...
+        setParameter(model.inner, string(name), string(value))
```
Comment (mlubin) [inline src/MOIWrapper.jl]: What happens if an invalid option name is passed?

File: src/MOIWrapper.jl (hunk 1 above)
Comment (vitornesello) [reply src/MOIWrapper.jl]: To differentiate the default constructor to the one with kwargs.. I took CPLEX.jl example. Apparently I cannot define a default constructor that takes kwargs.

File: src/MOIWrapper.jl (hunk 3 above)
Comment (vitornesello) [reply src/MOIWrapper.jl]: Cbc ignores unknows names

File: src/MOIWrapper.jl (hunk 1 above)
Comment (vitornesello) [reply src/MOIWrapper.jl]: But then tests fail for julia 0.6

Comment (mlubin) [reply src/MOIWrapper.jl]: Could you remove the inner constructor?

Then it should be valid to call `Optimizer(CbcModel())` from the outer constructor that takes keyword arguments.

File: src/MOIWrapper.jl (hunk 3 above)
Comment (mlubin) [reply src/MOIWrapper.jl]: Could you add a comment here about that?

File: src/MOIWrapper.jl (hunk 1 above)
Comment (vitornesello) [reply src/MOIWrapper.jl]: Yes, you are right. I did it

**Comment by vitornesello** [comment]
I am struggling to add the tests. Using cachingoptimizer to setup the model causes MOI to eventually call the `MOI.empty(::MOI.Optimizer)` function, which erases all the parameters.

**Comment by mlubin** [comment]
The implementation of `MOI.empty!` isn't supposed to erase the parameters.

http://www.juliaopt.org/MathOptInterface.jl/dev/apireference/#MathOptInterface.empty!
> Empty the model, that is, remove all variables, constraints and model attributes but not optimizer attributes.

Solver-specific parameters should be handled like optimizer attributes.

**Comment by vitornesello** [comment]
Ok, I will re-implement the function `empty`

**Comment by mlubin** [comment]
Bump. This would be nice to have ready for the 0.19 release.

**Pull request closed without merging**